### ISSUE TITLE
`TestBinaryErosionAndDilation` is flaky

### ISSUE THREAD

**Issue by emcastillo** [cupy/cupy#4021]
https://jenkins.preferred.jp/job/chainer/job/cupy_pr/2163/#showFailuresLink

```
tests.cupyx_tests.scipy_tests.ndimage_tests.test_morphology.BinaryErosionAndDilation_param_123_{border_value=0, connectivity=1, density=0.1, filter='binary_dilation', iterations=0, origin=0, output=None, shape=(16, 15), x_dtype=float32}.test_binary_erosion_and_dilation
--
tests.cupyx_tests.scipy_tests.ndimage_tests.test_morphology.BinaryErosionAndDilation_param_269_{border_value=0, connectivity=1, density=0.5, filter='binary_dilation', iterations=0, origin=0, output=None, shape=(5, 7, 9), x_dtype=float32}.test_binary_erosion_and_dilation
--
```

**Comment by grlee77** [comment]
I did see this on a couple of other PRs and it was a different parameterized case each time. I haven't been able to reproduce it locally. It does seem like it has always been the `cupy-py3` configuration that failed, though.


This was the case that came up once in the greyscale morphology PR:

`
tests.cupyx_tests.scipy_tests.ndimage_tests.test_morphology.BinaryErosionAndDilation_param_279_{border_value=0, connectivity=1, density=0.5, filter='binary_dilation', iterations=0, origin=-1, output=None, shape=(16, 15), x_dtype=float32}.test_binary_erosion_and_dilation
`

**Comment by emcastillo** [comment]
Maybe is the scipy version in cupy-py3

**Comment by grlee77** [comment]
One common aspect of the above is the fact that "iterations=0" which is a more complicated code path. In that case, the filter is applied iteratively until there is no change from the prior iteration.

**Comment by emcastillo** [comment]
I think it is reasonable to skip the test based on iteration value and scipy version

**Comment by toslunar** [comment]
I found a comparison with `empty_like`, which may be a cause of the flakiness.
https://github.com/cupy/cupy/blob/f94101b57a1c3231e5c5b97ba2b8e6787fbf57e5/cupyx/scipy/ndimage/morphology.py#L222-L231

**Comment by grlee77** [comment]
> I found a comparison with empty_like, which may be a cause of the flakiness.

Yeah, that looks pretty suspicious. I think we can just replace the last line there with:

```Python
changed = not (input == tmp_out).all()
```

Let me know if you intend to make a PR for that, otherwise I can submit one later today.

In SciPy, the erosion [function call returns a boolean indicating whether any values were changed](https://github.com/scipy/scipy/blob/a240c178a3f80461d9aef7daf3d4d9de5bdbdfcb/scipy/ndimage/morphology.py#L277-L279). I didn't see a great way to do the equivalent from a CUDA kernel, though, so have just done this check afterward instead. One could have a 0-dim array for `changed` passed into the kernel, but then this would have to be checked and updated with an atomic operation when a value is changed, which I thought would probably cause substantial overhead. It would also still have to be transferred back to the host for it to be used in boolean statements in the loop, so I don't think we could avoid synchronization that way either. I am interested if there might be another approach that I overlooked, though.
